### DISCUSSION TITLE
Struct vs list of struct

### DISCUSSION THREAD

**Discussion by jonas-eschle** [hep-statistics-standard/hep-statistics-standard#25] [Technical]
Should we use structs or list of structs?

**Comment by nsmith-** [comment]
If there is a desire to have a JSON schema which validation tools can work with, I think you'll have to go with list of structs, so that the struct keys are static and not dynamic. Perhaps there are options to extend that.

**Comment by matthewfeickert** [comment]
It would be helpful if @lukasheinrich could reiterate his thoughts from the last meeting here.

**Comment by lukasheinrich** [comment]
Hi - 

yes as @nsmith- says, for validation purposes it's good if keys are static as a part of a schema and values are variable. This was the reason why we changed this in pyhf originally from "channel_name: value" to [name: channel_name, data: value]

**Comment by lmoureaux** [comment]
> for validation purposes it's good if keys are static as a part of a schema and values are variable

Assuming JSON Schema, isn't it a matter of doing:
```json
{ "type": "object", "additionalProperties": { "type": "number" }}
```
instead of:
```json
{ "type": "array", "items": { "type": "number" }}
```
?

**Comment by lukasheinrich** [comment]
Hi @lmoureaux 

I think it's more about

```json
{
  "somename1": {"key":  "value" },
  "somename2": {"key":  "value" }
}
```

vs 

```json
[
  { "name": "somename1",
     "details":  {"key":  "value" }
   },
  { "name": "somename2",
     "details":  {"key":  "value" }
   },
]
```

**Reply by lmoureaux** [reply]
Sorry to come back to this after such a long time. Validation wise the two options are written pretty much the same in JSON Schema:
```json
{ "type": "object",
  "additionalProperties": { 
    "type": "object",
    "properties": {
      "key": {"type": "string"}
    }
  }
}
```
vs
```json
{ "type": "array",
  "items": { 
    "type": "object",
    "properties": {
      "name": {"type": "string"},
      "details": {
        "type": "object",
        "properties": {
          "key": {"type": "string"}
        }
      }
    }
  }
}
```
The additional complexity in the second schema comes from the nested `details` object which is not necessary in the first place. This demonstrates that the two options are similarly easy to validate.

**Comment by cburgard** [comment]
The current ROOT implementation supports both at the flick of a switch, but in order to avoid confusion, we should definitely converge on one of the two.
I personally have a preference for "structs" (that is, @lukasheinrich version 1) over "lists of structs (that is, @lukasheinrich version 2), just because it makes it easier to navigate/search things when looking at these files using e.g. a web browser, but then again, I don't know how much more difficult it would be to write validators for version 2 rather than version1.
More opinions would be highly welcome!

**Comment by alexander-held** [comment]
I am in favor of list-of-structs. I think it was @kratsg who convinced me that this is much better for validating and static keys are nice. Even if sometimes that requires an extra line of code, it can overall be a bit more readable as well.

**Reply by jonas-eschle** [reply]
I tend to agree, it's just cleaner and makes every object beautifully self-contained

**Comment by cburgard** [comment]
I'm not 100% convinced, I have to say.

On the readability-side, I'm quite convinced that dicts are much better than lists, since when using a GUI to view the file, you can directly see what an item is without having to expand it.
I'm also not sure I understand the point about the self-containedness. How is an object less self-contained if it is a dict?

And as far as the validation goes: I was a bit curious if validating is really that much harder when using arbitrary keys, and I spent ~2 hours writing a json schema to validate the dict-version of the standard. Please find it attached.
This is really just a first draft and all, but I'm not convinced that the ease of validation is really an argument in this case.
Perhaps you (or @kratsg) can make an argument about how/when validation fails with such a syntax, where it wouldn't in the list-version?
[schema.json.txt](https://github.com/hep-statistics-serialization-standard/hep-statistics-serialization-standard/files/12453083/schema.json.txt)

**Reply by alexander-held** [reply]
I agree about the GUI point, and about directly finding an entry by name in a dict. That is convenient. It probably is not a big runtime concern to have to search for it, implementations could internally re-convert if needed.

For self-containedness: to me this is about receiving a single object when iterating:
```python
for name, value in some_dict.items():
    ...
```
vs
```python
for value in some_list:
    ...
```
As you might not always care about the name in particular, the list-of-dicts treats it on the same footing as the rest.

**Reply by lmoureaux** [reply]
You also need to consider the use case where you want to look up something by name (because the very purpose of the name in the schema is to enable cross references):
```py
return objects[name]  # executes in O(1)
```
vs
```py
candidates = list(filter(lambda obj: obj['name'] == name, objects))  # executes in O(N) unless optimized explicitly to use a dict
if not candidates:
    raise ValueError(name)
elif len(candidates) > 1:
    raise ValueError(f'{name} not uniquely defined')
return candidates[0]
```
...unless you push the burden of checking `name` uniqueness to the validator, but then it's definitely easier to validate the "struct" solution.

**Comment by egpbos** [comment]
Does everything have a name? If so, it seems to me like the list-of-structs approach makes the JSON unnecessarily verbose. If not everything has a name, then the dict approach would be problematic, since one would have to artificially come up with names.

Another question: is the JSON meant to be used as-is in any place or will it always be converted to some kind of internal format (like a RooFit workspace)? If it is meant to be used directly, then the dict approach seems more convenient to me, since it makes it a lot easier to access directly (by key) specific components, otherwise a linear search would have to be done over all list elements.

Note: I have no strong opinions here, just some ideas that crossed my mind.

**Reply by egpbos** [reply]
By the way, should it be possible for elements to have duplicate names? If not, in the dict case, this is handled automatically, since there would be key clashes. In the list case, you would have to check all names against each other ($O(N)$ when checking this on adding elements).

**Reply by cburgard** [reply]
Thanks a lot for your input!

Yes, in the `distributions`, `functions` and `data` sections, which are the most relevant here, everything has a unique name for sure.
Individual types of functions or distributions, such as the histfactory construction, internally have other lists of things, such as the list of modifiers of the histfactory sample. These are not unique, since there can be multiple modifiers connected to the same parameter (think OverallSys+HistoSys), so in that case, we'd probably stick with the list-approach. But this is then about the specific features of how an individual distribution is represented (in this case, HistFactory), not so much about the standard itself.

**Reply by alexander-held** [reply]
Another thought on name uniqueness: I could imagine wanting to pack a large number of pdfs for possible signal models into the same workspace. Those could be turned on one-at-a-time to test different hypotheses. It might be slightly cleaner to distinguish them not by name (which would need to be extended to include all the possible settings of parameters that make each signal unique, and consumers would have to parse that string) but instead by additional keys that describe the signal (like `mass` or `angle` and such) as key-value pairs inside a dict (which would be a point in favor of list-of-dicts).

**Reply by egpbos** [reply]
You can still do the latter with key-value pairs (just discard the keys). But to be honest, when complex selection criteria are a serious desideratum, I think the way to go would be a serious database engine, not a JSON file :)

**Reply by alexander-held** [reply]
I'm not sure what you mean by discarding the keys. What I had in mind is something like this:
```yaml
{
  "signal_mass500_alpha0.7": {...},
  ...
}
```
vs
```yaml
[
  {
    "name": "signal",
    "mass": 500,
    "alpha": 0.7,
    ...
  },
  {...}
]
```
As far as I can tell, the former approach forces me to ensure my keys are unique and will end up with some custom strings that pack extra information in them that will either duplicate what is stored in the corresponding value or need to be parsed to extract the information.

I am not sure if non-unique names would end up causing other issues elsewhere though.

**Reply by egpbos** [reply]
Ah, thanks for the clarification. I was thinking instead about how one would have to search through the list to find the pdf matching the desired parameters. A unique key on a single parameter allows for more convenient access than having to filter on that single parameter. But, indeed, when you need to filter for other properties than the name, then a key won't help you either (that's what I meant by "discarding the key"), you'll still have to look at the values.

**Reply by lmoureaux** [reply]
I think @alexander-held's point is interesting but supporting it would complicate things tremendously. Just imagine the syntax for referencing an object with such complex keys in `function`. Consider the arguments in #12 (and #14). If an advanced indexing feature like this is desired as part of the first version of the standard, I think we first need a draft of how it would work with other parts of the current draft.

**Comment by kratsg** [comment]
Nobody really seems to be threading in this discussion which makes things so unnnecessarily burdensome for me to follow. If we're going to move discussions over to here from email threads, it would be great if people could follow the etiquette to make it easier to track the various comments.

First, a list of structs is wildly better than a struct of structs. When it comes to cross-language support, it is typically easier and more intuitive if -- given the sub-structs have same information -- for them to be collected in a list rather than in a struct. When someone provides a list of data points, they don't expect that data to be a struct of values, but a list.

```
"data": {"SR_A": 10.0, "SR_B": 12.0}
```

is completely weird to deal with programmatically but it's human-readable -- versus

```
"data": [{"name": "SR_A", "value": 10.0}, {"name": "SR_B", "value": 12.0}]
```

The latter approach is significantly better for a few reasons.
- the name of the channels (in this case) are user-defined, and are not fixed which means a schema needs to be defined allowing for keys following a particular pattern to ensure it all works well
- patching this requires that the patches include the custom defined names, which makes patches more fragile and susceptible to being changed more often. With the current approach of the HiFa JSON supported by pyhf, renaming channels or samples does not require changing the patch at all (to some extent)
- as pointed above by @lukasheinrich, static keys are just easier
- accessing the information programmatically will be inconsistent across languages, and many similarly-schema-implemented libraries have run into the "escaping-the-name" problem, where they try to provide a nice object for accessing the information in these things

Here's another fun example, which should be allowable and not break anyone's code

```
"data": {"SR/A": 10.0, "SR-B": 12.0, "SR.C": 14.0}
```

Think of trying to access these regions. `jq` will struggle with the `SR-B` and `SR.C` variations (as well as javascript), `ROOT` will struggle with `SR/A`, `python` will struggle with all three.

`model.data.SR/A`, `model.data.SR-B`, and `model.data.SR.C` all will fail to work using dot-access notation, and one is forced to go via `model.data["SR/A"]`, `model.data["SR-B"]`, and `model.data["SR.C"]` instead. If this was written as a list-of-structs instead:

```
"data": [{"name": "SR/A", "value": 10.0}, {"name": "SR-B", "value": 12.0}, {"name": "SR.C",  "value": 14.0}]
```

then `model.data.#.name` works (I'm using GJSON notation here) or `/model/data/*/name` (XPath notation which is used for JSON Patch) without having to think about what's going on here. It also forces the developers on top of these schema not to treat the values as being necessarily "simple" or following any particular set of rules that might be restrictive or imposed by the language they're programming in.

You need to pick and choose what your goal here is with the schema as you're mixing up different priorities. The choice of a JSON or plaintext representation is to make it more obvious to someone what the probability model looks like without being forced to install ROOT or another library to dump that information out. The XML had the problem of serializing the data separately in a binary format with an enforced structure that was tied to how it was defined in the XML (which is not great).

The point is that by going to plaintext, you want to be verbose. Arguing that a struct-of-structs is better than a list-of-structs because it's less verbose is a red herring and loses sight of the original point of moving the schema to something plain-text, serializable, and library-independent.

Second, using a GUI to view a file is also a strawman argument. Anyone can write a custom wrapper around rich to re-render a JSON that follows a schema into something more human-readable. This is what numpy and pandas already does. It doesn't render the entire tensor into view, just a subset of it. If people want more information, they have access to do so, but you're already going to be rendering the view differently anyway.

**Reply by nsmith-** [reply]
Regarding threading, I think it is just because this was moved from an issue to a discussion. So at the start we were not threading.

**Reply by nsmith-** [reply]
I think your point
> The choice of a JSON or plaintext representation is to make it more obvious to someone what the probability model looks like without being forced to install ROOT or another library to dump that information out. The XML had the problem of serializing the data separately in a binary format with an enforced structure that was tied to how it was defined in the XML (which is not great).

is excellent. We're defining a data format, correct? This means machine ergonomics are more important than human ones. I don't think we expect most (any?) users to hand-write the JSON.

**Reply by kratsg** [reply]
However, we're not talking about JSON but about schema. One can define the schema using JSON Schema, but the actual data format is an open choice separate from the schema choices itself. This latter part is a similar discussion to what is happening with `hist` / `boost-histogram` as well. One could use hdf and still have the same schema to validate against. In pyhf, we allow not just lists, but [also tensors](https://github.com/scikit-hep/pyhf/blob/c096b33cf859246771e06099ee1b48063ca051cd/src/pyhf/schema/validator.py#L84-L88), as part of the schema.

**Reply by lmoureaux** [reply]
Let me try to reply to a few things here (not the ones saying that something is `just better`, as this is just someone's opinion):

> When someone provides a list of data points, they don't expect that data to be a struct of values, but a list.

This is a circular argument, if you assume a list then you'll want to store it as a list. If you see it as a mapping between names and values instead, then an object is adequate. Since everything in the standard is referenced by name, I think the mapping option is more natural.

> ```
> "data": {"SR_A": 10.0, "SR_B": 12.0}
> ```
> is completely weird to deal with programmatically -- versus
> ```
> "data": [{"name": "SR_A", "value": 10.0}, {"name": "SR_B", "value": 12.0}]
> ```

I'm curious to see examples of what you mean. Python and C++ support both options equally well for iteration (with the obvious small syntax differences), and have an easier time with the first option for accessing an element by name -- which is needed to make sense of the file at all so will be a quite common use case.

> * the name of the channels (in this case) are user-defined, and are not fixed which means a schema needs to be defined allowing for keys following a particular pattern to ensure it all works well

This is #12/#24. In a JSON Schema, limiting the set of allowable keys is handled via `patternProperties` for the "struct" variant and `pattern` for the "list" variant. The schemas for either case will be very similar.

> * accessing the information programmatically will be inconsistent across languages, and many similarly-schema-implemented libraries have run into the "escaping-the-name" problem, where they try to provide a nice object for accessing the information in these things

Maybe it's just me, but I fail to see how escaping is made easier by encoding the name differently. In particular, the name is still supposed to be unique even in the "list" option since it's used as a key for object cross-references everywhere in the schema.

> Here's another fun example, which should be allowable and not break anyone's code
> ```
> "data": {"SR/A": 10.0, "SR-B": 12.0, "SR.C": 14.0}
> ```
> Think of trying to access these regions. jq will struggle with the SR-B and SR.C variations (as well as javascript), ROOT will struggle with SR/A, python will struggle with all three.

This is #12/#24 again. The keys you used would all break `generic_dist`. Note that the same issues with weird names and `generic_dist` happen regardless of how names are stored.
